File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/sync/library/items/files/system/MediaFileIdProvider.java
package com.lasthopesoftware.bluewater.client.sync.library.items.files.system;

import android.database.Cursor;
import android.provider.MediaStore;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.storage.read.permissions.IStorageReadPermissionArbitratorForOs;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.response.ImmediateResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Created by david on 6/13/16.
 */
public class MediaFileIdProvider implements ImmediateResponse<Cursor, Integer> {

	private static final Logger logger = LoggerFactory.getLogger(MediaFileIdProvider.class);
	private static final String audioIdKey = MediaStore.Audio.keyFor("audio_id");

	private final IMediaQueryCursorProvider mediaQueryCursorProvider;
	private final IStorageReadPermissionArbitratorForOs externalStorageReadPermissionsArbitrator;

	public MediaFileIdProvider(IMediaQueryCursorProvider mediaQueryCursorProvider, IStorageReadPermissionArbitratorForOs externalStorageReadPermissionsArbitrator) {
		this.mediaQueryCursorProvider = mediaQueryCursorProvider;
		this.externalStorageReadPermissionsArbitrator = externalStorageReadPermissionsArbitrator;
	}

	public Promise<Integer> getMediaId(ServiceFile serviceFile) {
		if (!externalStorageReadPermissionsArbitrator.isReadPermissionGranted())
			return new Promise<>(-1);

		return
			mediaQueryCursorProvider
				.getMediaQueryCursor(serviceFile)
				.then(this);
	}

	@Override
	public Integer respond(Cursor cursor) {
		if (cursor == null) return -1;

		try {
			if (cursor.moveToFirst())
				return cursor.getInt(cursor.getColumnIndexOrThrow(audioIdKey));
		} catch (IllegalArgumentException ie) {
			logger.info("Illegal column name.", ie);
		} finally {
			cursor.close();
		}

		return -1;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/system/MediaFileIdProvider.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.system

import android.database.Cursor
import android.provider.MediaStore
import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.shared.android.permissions.CheckOsPermissions
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.response.ImmediateResponse

class MediaFileIdProvider(
	private val mediaQueryCursorProvider: IMediaQueryCursorProvider,
	private val externalStorageReadPermissionsArbitrator: CheckOsPermissions
) : ImmediateResponse<Cursor?, Int>, ProvideMediaFileIds {
	override fun getMediaId(libraryId: LibraryId, serviceFile: ServiceFile): Promise<Int> =
		if (externalStorageReadPermissionsArbitrator.run { !isReadPermissionGranted || !isReadMediaAudioPermissionGranted }) Promise(-1)
		else mediaQueryCursorProvider
			.getMediaQueryCursor(libraryId, serviceFile)
			.then(this)

	override fun respond(cursor: Cursor?): Int {
		if (cursor == null) return -1
		try {
			if (cursor.moveToFirst()) return cursor.getInt(cursor.getColumnIndexOrThrow(audioIdKey))
		} catch (ie: IllegalArgumentException) {
			logger.info("Illegal column name.", ie)
		} finally {
			cursor.close()
		}
		return -1
	}

	companion object {
		private val logger by lazyLogger<MediaFileIdProvider>()
		private const val audioIdKey = MediaStore.Audio.Media._ID
	}
}
